# Data generation for the classification problem

In [24]:
# Import libraries
import pandas as pd
from sklearn.datasets import make_classification
from sklearn.model_selection import train_test_split

In [25]:
# Define data for make_classification()

samples = 220
features = 14
informative = 12
bias = 2
noise = 50
seed = 2224425784

In [26]:
# Create data in X, y from make_clasdification
X, y = make_classification(n_samples=samples, n_features=features, n_informative=informative, n_redundant=bias, shift=noise, shuffle=False, random_state=seed)

In [27]:
# Show info dataframe
df = pd.DataFrame(X,y)
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 220 entries, 0 to 1
Data columns (total 14 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   0       220 non-null    float64
 1   1       220 non-null    float64
 2   2       220 non-null    float64
 3   3       220 non-null    float64
 4   4       220 non-null    float64
 5   5       220 non-null    float64
 6   6       220 non-null    float64
 7   7       220 non-null    float64
 8   8       220 non-null    float64
 9   9       220 non-null    float64
 10  10      220 non-null    float64
 11  11      220 non-null    float64
 12  12      220 non-null    float64
 13  13      220 non-null    float64
dtypes: float64(14)
memory usage: 25.8 KB


When creating the df data frame, 220 entries were created in 14 columns, this data can vary when defining new data in parameters, samples, characteristics, informational, etc.

In [28]:
# Split data 
X_train , X_test , y_train , y_test = train_test_split(X,y, test_size =0.09, shuffle=True, random_state=3, stratify=y)

The data required for the test is 20 and the training data is 200, for this, the parameters used for this specific case are the test size = 0.09, the value of 0.09
comes from the relationship between the observations required for training and the observations
tions of the test that is, 20/200 = 0.1, making a small adjustment we have the value of 0.09,
Additionally, we place the parameter stratify = y, so that the random return
obtain the appropriate percentage of 0 and 1 corresponding to the categorical variable (0, 1), printing the values on the screen we can verify that the changes have indeed been
done.

In [29]:
print(len(X_train), 'X_train')
print(len(y_train), 'y_train')
print(len(X_test), 'X_test')
print(len(y_test), 'y_test')


200 X_train
200 y_train
20 X_test
20 y_test


In [30]:
# Description of training data and test data
dftrain = pd.DataFrame(X_train ,y_train) 
dftrain.info()
#dftrain.describe()
dftest = pd.DataFrame(X_test ,y_test)
dftest.info() 
#dftest.describe()

<class 'pandas.core.frame.DataFrame'>
Index: 200 entries, 1 to 0
Data columns (total 14 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   0       200 non-null    float64
 1   1       200 non-null    float64
 2   2       200 non-null    float64
 3   3       200 non-null    float64
 4   4       200 non-null    float64
 5   5       200 non-null    float64
 6   6       200 non-null    float64
 7   7       200 non-null    float64
 8   8       200 non-null    float64
 9   9       200 non-null    float64
 10  10      200 non-null    float64
 11  11      200 non-null    float64
 12  12      200 non-null    float64
 13  13      200 non-null    float64
dtypes: float64(14)
memory usage: 23.4 KB
<class 'pandas.core.frame.DataFrame'>
Index: 20 entries, 1 to 0
Data columns (total 14 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   0       20 non-null     float64
 1   1       20 non-null     float64
 2   2       20 non-nul

# Assembly models

In [31]:
# Import libraries
from sklearn.ensemble import BaggingClassifier
from sklearn.tree import DecisionTreeClassifier

## BagginClassifier

In [32]:
bag_clf = BaggingClassifier(
    DecisionTreeClassifier(random_state = 3), n_estimators =500,
    max_samples = 75, bootstrap=True, random_state = 3)

### Adjust Model

In [33]:
bag_clf.fit(X_train, y_train)
y_pred_bc = bag_clf.predict(X_test)

### Result

In [44]:
from sklearn.metrics import accuracy_score
pbc = accuracy_score(y_test, y_pred_bc)
print(pbc)

0.75


## Pasting

In [45]:
pas_clf = BaggingClassifier (
DecisionTreeClassifier ( random_state = 3) , n_estimators = 500 ,
max_samples = 75 , bootstrap = False , random_state = 3)

### Adjust Model

In [46]:
pas_clf.fit ( X_train , y_train )

BaggingClassifier(bootstrap=False,
                  estimator=DecisionTreeClassifier(random_state=3),
                  max_samples=75, n_estimators=500, random_state=3)

In [53]:
y_pred_pas = pas_clf.predict ( X_test )
ppas = accuracy_score(y_test , y_pred_pas)
print (ppas)

0.75


## Assembly model Ramdom Forest

In [48]:
from sklearn . ensemble import RandomForestClassifier
rnd_clf = RandomForestClassifier ( n_estimators = 500 , max_leaf_nodes =
4 , random_state = 3 , max_samples = 75)

### Adjust model

In [55]:
rnd_clf.fit(X_train,y_train)
y_pred_rf = rnd_clf . predict ( X_test )
prf = accuracy_score(y_test,y_pred_rf)
print (prf)

0.75


## Assembly model Gradient Boosting

In [50]:
from sklearn.ensemble import GradientBoostingClassifier
gb_clf = GradientBoostingClassifier ( n_estimators = 500 , random_state= 3)

### Adjust model

In [57]:
gb_clf.fit (X_train,y_train ) 
y_pred_gb = gb_clf.predict ( X_test ) 
pboo = accuracy_score(y_test,y_pred_gb)
print(pboo)

0.85


## Analysis

In [66]:
df = pd.DataFrame({"y_test": y_test,"y_pred_bc": y_pred_bc, "y_pred_pas":y_pred_pas, "y_pred_rf":y_pred_rf,"y_pred_gb":y_pred_gb})
print()
print(' y_test     = Data Testing\n',
      'y_pred_bc  = Baggin Classifier ---> Accuracy '+str(pbc)+"\n",
      'y_pred_pas = Pasting           ---> Accuracy '+str(ppas)+"\n",
      'y_pred_rf  = Random Forest     ---> Accuracy '+str(prf)+"\n",
      'y_pred_gb  = Gradient Boosting ---> Accuracy '+str(pboo)
     )
df.head(15)


 y_test     = Data Testing
 y_pred_bc  = Baggin Classifier ---> Accuracy 0.75
 y_pred_pas = Pasting           ---> Accuracy 0.75
 y_pred_rf  = Random Forest     ---> Accuracy 0.75
 y_pred_gb  = Gradient Boosting ---> Accuracy 0.85


y_test  y_pred_bc  y_pred_pas  y_pred_rf  y_pred_gb
0        1          1           1          1          1
1        0          0           0          1          0
2        0          0           0          0          0
3        0          0           0          1          0
4        1          1           1          1          1
5        1          1           1          1          1
6        0          1           1          0          1
7        1          0           0          0          1
8        0          1           1          1          1
9        1          1           1          1          1
10       0          0           0          0          0
11       0          0           0          0          0
12       1          1           1          1          1
13       0          0           0          0          0
14       1          1           1          1          1

The best result for the created data frame is the Gradient Boosting ensemble model with 0.85 of hits